python/taichi/ad/_ad.py

```diff
@@ -205,24 +206,32 @@ def __enter__(self):
 
         # Attach the context manager to runtime
         self.runtime.target_tape = self
+        return self
 
     def __exit__(self, _type, value, tb):
         self.runtime.target_tape = None
         if self.eval_on_exit:
             self.grad()
+        for calls, mode in zip(self.calls, self.modes):
+            calls[0].autodiff_mode = mode
 
     def insert(self, func, args):
+        if self.validation and not impl.get_runtime().grad_replaced:
+            self.modes.append(func.autodiff_mode)
```

Comment:
this isn't safe since `self.modes` could be a different length from `self.calls` which breaks your `zip` in `__exit__`.